File: java/com/aa/act/interview/org/Organization.java
package com.aa.act.interview.org;

import java.util.Optional;
import java.util.Queue;
import java.util.LinkedList;

public abstract class Organization {

    private Position root;
    
    public Organization() {
        root = createOrganization();
    }
    
    protected abstract Position createOrganization();
    
    /**
     * hire the given person as an employee in the position that has that title
     * 
     * @param person
     * @param title
     * @return the newly filled position or empty if no position has that title
     */
    public Optional<Position> hire(Name person, String title) {
        //your code here
        Queue<Position> queue = new LinkedList<Position>();
        queue.add(root);
        int identifierCount = 1;
        while (!queue.isEmpty()) {
            Position pos = queue.remove();
            if (pos.getTitle().equals(title)) {
                pos.setEmployee(Optional.of(new Employee(identifierCount, person)));
                return Optional.of(pos);
            }
            identifierCount++;
            queue.addAll(pos.getDirectReports());
        }
        return Optional.empty();
    }

    @Override
    public String toString() {
        return printOrganization(root, "");
    }
    
    private String printOrganization(Position pos, String prefix) {
        StringBuffer sb = new StringBuffer(prefix + "+-" + pos.toString() + "\n");
        for(Position p : pos.getDirectReports()) {
            sb.append(printOrganization(p, prefix + "  "));
        }
        return sb.toString();
    }
}

File: javascript-es6/Organization.js
class Organization {
  constructor(root) {
    this.printOrganization = (position, prefix) => {
      let str = `${prefix}+-${position.toString()}\n`;
      for (const p of position.getDirectReports()) {
        str = str.concat(this.printOrganization(p, `${prefix}  `));
      }
      return str;
    };

    // Hire the given person as an employee in the position that has that title
    // Return the newly filled position or undefined if no position has that title
    this.hire = (person, title) => {
      // your code here
      var queue = [root];
      while (queue.length) {
        var position = queue.shift();
        if (position.getTitle() === title) {
          position.setEmployee(person);
          return position;
        }
        queue.push(...position.getDirectReports());
      }
      return undefined;
    };

    this.toString = () => this.printOrganization(root, '');
  };
}

export default Organization;
